catwalk/steps.py

```diff
@@ -257,6 +260,8 @@ def run(
             state = torch.load(train_config.final_weights_path, map_location="cpu")
             # We use `strict=False` because there might be missing keys due to weight tying.
             trainable_model.load_state_dict(state, strict=False)
+            # if distributed, this model hasn't been sent to device yet
+            trainable_model.to(resolve_device())
```

Comment:
Why do we have to return a model that's tied to a device?

Reply:
> Why do we have to return a model that's tied to a device?

Hmm I had a crash where not moving the model back to the GPU would cause inputs on the GPU to attempt to be processed in the CPU model during the final evaluation. I can't replicate this now thought. I suspect it was required for the failed attempt to integrate the fairscale code in #79. 

As it stands here there is no crash without this code, but it does mean that the final evaluation will run on CPU rather than GPU as it would it if it were running a non-trainable model. I can't just make the `TrainableRankClassificationModel.predict()` use `resolve_device()` like the non-trainable `predict()` does, because invoking predict at all inside the training process would mess up the distributed device map. 

Perhaps the best place to move it to GPU would be in `catwalk.train.py` like this?
```
model_step = FinetuneStep(
            model=args.model,
            tasks=tasks,
            batch_size=args.batch_size,
            grad_accum=args.grad_acc,
            device_count=args.device_count
        )

model_step = model_step.result().to(resolve_device())
```